src/api/focus.js

```diff
@@ -111,114 +111,112 @@ class Focus {
-        function: "waitForSerialDevice",
-      });
-      await delay(2000);
     }
-
-    logger("focus").warn("serial device not found after 10 attempts", {
-      function: "waitForSerialDevice",
+    logger("focus").debug("serial device not found", {
+      function: "checkForSerialDevice",
+      device: usbInfo,
     });
+
     return null;
   }
 
-  async waitForSerialBootloader(focusDeviceDescriptor) {
-    return await this.waitForSerialDevice(
+  async checkSerialBootloader(focusDeviceDescriptor) {
+    return await this.checkSerialDevice(
       focusDeviceDescriptor,
       focusDeviceDescriptor.usb.bootloader
     );
   }
 
-  async waitForDFUBootloader(focusDeviceDescriptor) {
-    const delay = (ms) => new Promise((res) => setTimeout(res, ms));
+  async checkNonSerialBootloader(focusDeviceDescriptor) {
     const bootloader = focusDeviceDescriptor.usb.bootloader;
 
-    for (let attempt = 0; attempt < 10; attempt++) {
-      const deviceList = await ipcRenderer.invoke(
-        "usb.scan-for-devices",
-        bootloader.productId,
-        bootloader.vendorId
-      );
-
-      for (const device of deviceList) {
-        const pid = device.deviceDescriptor.idProduct,
-          vid = device.deviceDescriptor.idVendor;
-
-        if (pid == bootloader.productId && vid == bootloader.vendorId) {
-          logger("focus").info("bootloader found", {
-            device: bootloader,
-            function: "waitForDFUBootloader",
-          });
-          return true;
-        }
-      }
+    const deviceList = await ipcRenderer.invoke(
+      "usb.scan-for-devices",
+      bootloader.productId,
+      bootloader.vendorId
+    );
 
-      logger("focus").debug("bootloader not found, waiting 2s", {
-        function: "waitForDFUBootloader",
-      });
-      await delay(2000);
+    for (const device of deviceList) {
+      const pid = device.deviceDescriptor.idProduct,
+        vid = device.deviceDescriptor.idVendor;
+
+      if (pid == bootloader.productId && vid == bootloader.vendorId) {
+        const newPort = Object.assign({}, device);
+        newPort.focusDeviceDescriptor = focusDeviceDescriptor;
+        newPort.focusDeviceDescriptor.bootloader = true;
+
+        logger("focus").info("bootloader found", {
+          device: bootloader,
+          function: "checkNonSerialBootloader",
+        });
+        return newPort;
+      }
     }
 
-    logger("focus").warn("bootloader not found after 10 attempts", {
-      function: "waitForDFUBootloader",
+    logger("focus").debug("bootloader not found", {
+      function: "checkNonSerialBootloader",
+      device: bootloader,
     });
-    return false;
+
+    return null;
   }
 
-  async waitForBootloader(focusDeviceDescriptor) {
+  async checkBootloader(focusDeviceDescriptor) {
     if (!focusDeviceDescriptor.usb.bootloader) {
-      this.logger.warn("No bootloader defined in the device descriptor", {
+      this.logger().warn("No bootloader defined in the device descriptor", {
```

Comment:
What's the difference between this.logger().warn and  logger("focus").info other than log level? we use both in this function

Reply:
The difference is that `logger()` exists, and `this.logger()` does not. =)